File: pluginbase-core/messages/src/test/java/pluginbase/messages/MessageTest.java
package pluginbase.messages;

import ninja.leaping.configurate.hocon.HoconConfigurationLoader;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.cglib.core.Local;
import pluginbase.logging.Logging;
import pluginbase.logging.PluginLogger;

import java.io.File;
import java.util.Arrays;
import java.util.Locale;

public class MessageTest {

    final String testKey = "this is my key";
    final String testMessage = "this is my default message: %s";

    enum TestEnumMessages implements Message, LocalizablePlugin {
        IN_GAME_ONLY("commands.in_game_only", "To use from console, use \"-p <name>\" to specify a player"),
        INVALID_AMOUNT("commands.invalid_amount", "'%s' is not a valid amount!"),
        ;

        private final Message message;

        TestEnumMessages(@NotNull String key, @NotNull String defaultMessage, String... additionalLines) {
            this.message = Message.createMessage(key, defaultMessage, additionalLines);
        }

        @Override
        @NotNull
        public String getDefault() {
            return message.getDefault();
        }

        @Override
        @Nullable
        public Object[] getKey() {
            return message.getKey();
        }

        @Override
        public int getArgCount() {
            return message.getArgCount();
        }

        @NotNull
        @Override
        public PluginLogger getLog() {
            return Logging.getLogger();
        }
    }

    @Before
    public void setUp() {
        // Tests the loading of theme.xml
        Messages.BLANK.getDefault();
    }

    @Test
    public void testEnumMessages() throws Exception {
        Messages.registerMessages(TestEnumMessages.IN_GAME_ONLY, TestEnumMessages.class);
        Assert.assertNotNull(Messages.getMessage(TestEnumMessages.IN_GAME_ONLY, TestEnumMessages.INVALID_AMOUNT.getKey()));
        MessageProvider provider = Messages.loadMessages(TestEnumMessages.IN_GAME_ONLY,
                HoconConfigurationLoader.builder().setFile(File.createTempFile("pluginbase", "test")).build(),
                Locale.ENGLISH);
        Assert.assertEquals(TestEnumMessages.IN_GAME_ONLY.getDefault(), provider.getLocalizedMessage(TestEnumMessages.IN_GAME_ONLY));
    }

    @Test
    public void testCreateMessage() throws Exception {
        final Message message = Message.createMessage(testKey, testMessage);
        Assert.assertEquals(Arrays.toString(new Object[] {testKey}), Arrays.toString(message.getKey()));
        Assert.assertEquals(testMessage, message.getDefault());
    }

    @Test
    public void testBundleMessage() throws Exception {
        final Message message = Message.createMessage(testKey, testMessage);
        //Message.bundleMessage()
    }
}

File: pluginbase-core/messages/src/test/kotlin/pluginbase/messages/MessageTest.kt
package pluginbase.messages

import ninja.leaping.configurate.hocon.HoconConfigurationLoader
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import pluginbase.logging.Logging
import pluginbase.logging.PluginLogger

import java.io.File
import java.util.Arrays
import java.util.Locale

class MessageTest {

    val testKey = "this is my key"
    val testMessage = "this is my default message: %s"

    internal enum class TestEnumMessages constructor(key: String, defaultMessage: String,
                                                     vararg additionalLines: String) : LocalizablePlugin, Message {
        IN_GAME_ONLY("commands.in_game_only", "To use from console, use \"-p <name>\" to specify a player"),
        INVALID_AMOUNT("commands.invalid_amount", "'%s' is not a valid amount!");

        val message: Message = Messages.createMessage(key, defaultMessage, *additionalLines)

        override fun getLog(): PluginLogger {
            return Logging.getLogger()
        }

        override val default: String
            get() = message.default
        override val key: Array<*>?
            get() = message.key
        override val argCount: Int
            get() = message.argCount
    }

    @Before
    fun setUp() {
        // Tests the loading of theme.xml
        Messages.BLANK.default
    }

    @Test
    @Throws(Exception::class)
    fun testEnumMessages() {
        Messages.registerMessages(TestEnumMessages.IN_GAME_ONLY, TestEnumMessages::class.java)
        Assert.assertNotNull(Messages.getMessage(TestEnumMessages.IN_GAME_ONLY, TestEnumMessages.INVALID_AMOUNT.message.key))
        val provider = Messages.loadMessages(TestEnumMessages.IN_GAME_ONLY,
                HoconConfigurationLoader.builder().setFile(File.createTempFile("pluginbase", "test")).build(),
                Locale.ENGLISH)
        Assert.assertEquals(TestEnumMessages.IN_GAME_ONLY.message.default, provider.getLocalizedMessage(TestEnumMessages.IN_GAME_ONLY.message))
    }

    @Test
    @Throws(Exception::class)
    fun testCreateMessage() {
        val message = Messages.createMessage(testKey, testMessage)
        Assert.assertEquals(Arrays.toString(arrayOf<Any>(testKey)), Arrays.toString(message.key))
        Assert.assertEquals(testMessage, message.default)
    }

    @Test
    @Throws(Exception::class)
    fun testBundleMessage() {
        val message = Messages.createMessage(testKey, testMessage)
        //Message.bundleMessage()
    }
}
